pintracker/stateless/stateless.go

```diff
@@ -0,0 +1,490 @@
+// If the Cid is pinned locally, it will be unpinned.
+func (spt *Tracker) Untrack(c *cid.Cid) error {
+	logger.Debugf("untracking %s", c)
+	return spt.enqueue(api.PinCid(c), optracker.OperationUnpin)
+}
+
+// StatusAll returns information for all Cids pinned to the local IPFS node.
+func (spt *Tracker) StatusAll() []api.PinInfo {
+	pininfos, err := spt.localStatus(true)
+	if err != nil {
+		logger.Error(err)
+		return nil
+	}
+
+	// get all inflight operations from optracker and
+	// put them into the map, deduplicating any already 'pinned' items with
+	// their inflight operation
+	for _, infop := range spt.optracker.GetAll() {
+		pininfos[infop.Cid.String()] = infop
+	}
+
+	var pis []api.PinInfo
+	for _, pi := range pininfos {
+		pis = append(pis, pi)
+	}
+	return pis
+}
+
+// Status returns information for a Cid pinned to the local IPFS node.
+func (spt *Tracker) Status(c *cid.Cid) api.PinInfo {
+	// check if c has an inflight operation or errorred operation in optracker
+	if oppi, ok := spt.optracker.GetExists(c); ok {
+		// if it does return the status of the operation
+		return oppi
+	}
+
+	// check global state to see if cluster should even be caring about
+	// the provided cid
+	var gpin api.PinSerial
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"PinGet",
+		api.PinCid(c).ToSerial(),
+		&gpin,
+	)
+	if err != nil {
+		if !rpc.IsRPCError(err) {
+			// if _, ok := err.(*api.CidNotInGlobalStateError); ok {
+			return api.PinInfo{}
+		}
+		logger.Error(err)
+	}
+
+	// check if pin is a remote pin
+	if gpin.ToPin().IsRemotePin(spt.peerID) {
+		return api.PinInfo{Cid: c, Peer: spt.peerID, Status: api.TrackerStatusRemote, TS: time.Now()}
+	}
+
+	// else attempt to get status from ipfs node
+	var ips api.IPFSPinStatus
+	err = spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"IPFSPinLsCid",
+		api.PinCid(c).ToSerial(),
+		&ips,
+	)
+	if err != nil {
+		logger.Error(err)
+		return api.PinInfo{}
+	}
+
+	pi := api.PinInfo{
+		Cid:    c,
+		Peer:   spt.peerID,
+		Status: ips.ToTrackerStatus(),
+	}
+
+	return pi
+}
+
+// SyncAll verifies that the statuses of all tracked Cids (from the shared state)
+// match the one reported by the IPFS daemon. If not, they will be transitioned
+// to PinError or UnpinError.
+//
+// SyncAll returns the list of local status for all tracked Cids which
+// were updated or have errors. Cids in error states can be recovered
+// with Recover().
+// An error is returned if we are unable to contact the IPFS daemon.
+func (spt *Tracker) SyncAll() ([]api.PinInfo, error) {
+	// get ipfs status for all
+	localpis, err := spt.localStatus(false)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationPin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationUnpin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
```

Comment:
Shouldn't it be `!ok`? If it's reported by IPFS, it's not been unpinned, thus it should not be Cleaned.

Reply:
Good catch, thanks.

Reply:
Means we need a test (probably missing in maptracker too?)